src/DSP/Reverter.h

```diff
@@ -0,0 +1,292 @@
+     */
+    void processBuffer(float *smp);
+
+    /**
+     * @brief Handle synchronization modes and state transitions.
+     * 
+     */
+    void handleSync();
+
+    /**
+     * @brief Handle synchronization in NOTEON and NOTEONOFF modes.
+     * 
+     */
+    void handleNoteSync();
+
+    /**
+     * @brief Handles state transitions in NOTEON and NOTEONOFF modes.
+     */
+    void handleStateChange();
+
+    /**
+     * @brief Update the read position in the ring buffer.
+     */
+    void updateReaderPosition(int i);
+
+    /**
+     * @brief Apply crossfading between two buffer segments.
+     * 
+     * @param smp Pointer to the buffer of input samples.
+     * @param i Index of the current sample in the buffer.
+     */
+    void crossfadeSamples(float *smp, int i);
+
+    /**
+     * @brief Apply fading for transitions between buffer segments.
+     * 
+     * @param fadein Fade-in factor.
+     * @param fadeout Fade-out factor.
+     * @return Crossfaded sample value.
+     */
+    float applyFade(float fadein, float fadeout);
+
+    /**
+     * @brief Apply the output gain to a sample.
+     * 
+     * @param sample Reference to the sample to modify.
+     */
+    void applyGain(float &sample);
+
+    /// Current synchronization mode
+    SyncMode syncMode;
+
+    /// Current state of the reverter (RECORDING, PLAYING, IDLE)
+    State state;
+
+    /// Ring buffer for input samples
+    float* input;
+
+    /// Output gain value
+    float gain;
+
+    /// Delay time in samples
+    float delay;
+
+    /// Phase value
+    float phase;
+
+    /// Crossfade duration in samples
+    float crossfade;
+
+    /// Time reference for syncing
+    float tRef;
+
+    /// Offset in the buffer for reading/writing
+    int buffer_offset;
+
+    /// Counter for buffer processing
+    int buffer_counter;
+
+    /// Offset for global time synchronization
+    float global_offset;
+
+    /// Index for reverse playback
+    float reverse_index;
+
+    /// Phase offset for delay transitions
+    float phase_offset_old;
+    float phase_offset_fade;
+
+    /// Number of samples for crossfading
+    int fading_samples;
+
+    /// Counter for crossfade processing
+    int fade_counter;
+
+    /// Root mean square (RMS) history for input volume analysis
+    float mean_abs_value;
+
+    /// Pointer to timing object for synchronization
+    const AbsTime *time;
+
+    /// Memory allocator
+    Allocator &memory;
+
+    /// Total size of the memory buffer in samples
+    unsigned int mem_size;
+
+    /// Sample rate of the audio engine
+    const int samplerate;
+
+    /// Size of the audio buffer in samples
+    const int buffersize;
+
+    /// Maximum allowable delay in samples
+    float max_delay;
+
+    /// Start position for the read head in the ring buffer
+    unsigned int pos_start;
+
+    /// Offset for phase correction
+    float phase_offset;
+
+    /// Number of samples recorded for playback
```

Comment:
While the reader can be positioned at non-integer values due to the phase offset (honestly rounding to the nearest sample might be fine there as well), shouldn't the number of recorded samples always be an integer? Not a huge problem since 24bits of float precision is plenty, just a thought while reading.

Reply:
actually it has always integer values, as does reverse_index.